deploy: pod deletion cost controller

szuecs opened the pull request:
it takes care of application=skipper-ingress,component=ingress to have a proper deletion cost as written down in https://github.com/kubernetes/kubernetes/issues/124149

szuecs committed:
deploy: pod deletion cost controller

it takes care of application=skipper-ingress,component=ingress to have a proper deletion cost as written down in https://github.com/kubernetes/kubernetes/issues/124149

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

AlexanderYastrebov commented on cluster/manifests/skipper/pod-deletion-cost-controller.yaml:
```diff
@@ -0,0 +1,58 @@
+  selector:
+    matchLabels:
+      deployment: pod-deletion-cost-controller
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 0
+    type: RollingUpdate
```
I think [Recreate strategy](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#recreate-deployment) fits better to ensure single instance like we use for fabric-gateway-controller

szuecs replied:
It should not matter and I do not think that "ensure single" is happening anyways, regardless of the config. 
Think of it more that "most often" it will be like you expect.
I will change as you suggest.

szuecs replied:
done

AlexanderYastrebov commented on cluster/manifests/skipper/pod-deletion-cost-controller.yaml:
```diff
@@ -0,0 +1,58 @@
+apiVersion: apps/v1
```
Do we want a feature toggle for it?

szuecs replied:
Why?
It does not harm to run it

szuecs committed:
pod del controller: add serviceaccount/rbac
fix: container use entrypoint so args[0] is already passed
fix: deployment strategy

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

AlexanderYastrebov commented:
:+1:

szuecs commented:
:+1:

rtomadpg commented:
If I understand the purpose of the "pod-deletion-cost-controller" correctly, it is the first of its kind.

Are there any plans for making it available outside Zalando? Much appreciated.

szuecs commented:
@rtomadpg first we have to check if it works as we expect, but normally I am one of the people that would OSS it. It's great to see if there are already folks asking for this so there is public demand for such a controller.
In the next weeks we will gather information and if it's good enough I have to polish the code such that backends will also get the pod deletion cost settings, right now it focuses only on our ingress implementation [skipper http proxy](https://opensource.zalando.com/skipper).

rtomadpg commented:
We're using Skipper, so 🚀 !

zhuoyang commented:
Any updates on this? I am interested to use this pod deletion cost controller to solve AZ balancing issue during scale down

szuecs commented:
> Any updates on this? I am interested to use this pod deletion cost controller to solve AZ balancing issue during scale down

I am on parental leave and I am not sure if the project is currently going on or paused until I come back. We will let you know when it is good enough to share. We got some more experience and will share it with the public later.
